### PR TITLE
binary: Handle invalid strings with large lengths

### PR DESCRIPTION
**Pull request by abhinav**
Currently, we always pre-allocate the buffer to read strings into. This means
that an invalid request with a really large string length can simply lock the
system up. For example, `0x22, 0x6e, 0x6f, 0x74` will request a string 500+
MB long. Even if that many bytes don't follow, we'll end up allocating that
much space.

This changes the system to use a dynamically resizing buffer if the requested
length is larger than 1 MB. This will cause invalid requests to fail gracefully
because that much data will not be available over the wire.

### CODE REVIEW

**Comment by abhinav** [comment]
CC @prashantv @breerly @bombela @kriskowal

File: protocol/binary_test.go [hunk 1]
```diff
@@ -284,6 +285,18 @@ func TestBinary(t *testing.T) {
...
+	// 5 MB + 4 bytes for length
+	data := make([]byte, 5242880+4)
+	data[0], data[1], data[2], data[3] = 0x0, 0x50, 0x0, 0x0 // 5 MB
+
+	value, err := Binary.Decode(bytes.NewReader(data), wire.TBinary)
+	require.NoError(t, err, "failed to decode value")
+
+	want := wire.NewValueBinary(data[4:])
+	assert.True(t, wire.ValuesAreEqual(want, value), "values did not match")
+}
```
Comment (prashantv) [inline protocol/binary_test.go]: this test will pass with or without this change right? i would have preferred a test where we have a small buffer but where the length is invalid (max size). However with 4 bytes, the length could be 4gb which may succeed with a make.

**Comment by prashantv** [comment]
lgtm, 👍 for fixing this. wish we could test this better, but unless there's something trivial, don't waste time on it.

File: protocol/binary_test.go (hunk 1 above)
Comment (abhinav) [inline protocol/binary_test.go]: The other test tests for invalid really-large length. I couldn't find a way to limit the amount of memory the test is allowed to allocate, so it'll still pass without this change.

File: protocol/binary_test.go [hunk 2]
```diff
@@ -297,6 +310,7 @@ func TestBinaryEOFFailure(t *testing.T) {
 		{},
 		{0x00}, // incomplete length
 		{0x00, 0x00, 0x00, 0x01}, // length mismatch
+		{0x22, 0x6e, 0x6f, 0x74}, // really long length
```
Comment (breerly) [inline protocol/binary_test.go]: Suggest moving comments into `desc` var that can be used during failure.

File: protocol/binary/reader.go [hunk 3]
```diff
@@ -202,6 +207,19 @@ func (br *Reader) read(bs []byte, off int64) (int64, error) {
 	return off, err
 }
 
+// copyN copies the given number of bytes starting at the given position into
+// the Writer.
+func (br *Reader) copyN(w io.Writer, off int64, n int64) (int64, error) {
```
Comment (breerly) [inline protocol/binary/reader.go]: I think we can name this better... `copyIntoWriter` perhaps?

Comment (breerly) [inline protocol/binary/reader.go]: Wow... just saw `io.CopyN`... Ok - I'm def wrong here :)

Comment (breerly) [inline protocol/binary/reader.go]: I think the comment on this function confused me. Suggest taking `io.CopyN`'s lead with something like:

```
// CopyN copies n bytes (or until an error) from ...
```

**Comment by breerly** [comment]
lgtm (with nits)

File: protocol/binary/reader.go (hunk 3 above)
Comment (abhinav) [inline protocol/binary/reader.go]: 💯

File: protocol/binary_test.go (hunk 2 above)
Comment (abhinav) [inline protocol/binary_test.go]: That will involve changing all `checkEOFError`-based tests. I would prefer to hold off on this change until this becomes an actual issue. Usually there are only 1-2 cases for each test based on that.

File: protocol/binary/reader.go (hunk 3 above)
Comment (abhinav) [inline protocol/binary/reader.go]: Forgot to push the change into this PR. Fix: https://github.com/thriftrw/thriftrw-go/pull/165

**Comment by blampe** [comment]
This is really great, how did you find this?

**Comment by abhinav** [comment]
A yarpc test was timing out when it received an invalid payload.

**Pull request merged**